source/extensions/clusters/dns/dns_cluster.cc

```diff
@@ -22,35 +21,264 @@ DnsClusterFactory::createClusterWithConfig(
+            }
+
+            // Note that this will trigger priority update callbacks everytime a logical
+            // DNS cluster receives a different address. In the previous implementation,
+            // it would only issue that callback the first time a DNS response was received.
+            // After that, it would change the address of the host "in place", without
+            // any changes to priorities and others. Now, as we're fully removing and adding
+            // hosts, we need to trigger the priority update callback every time.
+            parent_.updateAllHosts(hosts_added, hosts_removed, locality_lb_endpoints_.priority());
+          } else {
+            parent_.info_->configUpdateStats().update_no_rebuild_.inc();
+          }
+
+          // reset failure backoff strategy because there was a success.
+          parent_.failure_backoff_strategy_->reset();
+
+          if (!response.empty() && parent_.respect_dns_ttl_ &&
+              ttl_refresh_rate != std::chrono::seconds(0)) {
+            final_refresh_rate = ttl_refresh_rate;
+            ASSERT(ttl_refresh_rate != std::chrono::seconds::max() &&
+                   final_refresh_rate.count() > 0);
+          }
+          if (parent_.dns_jitter_ms_.count() > 0) {
+            // Note that `parent_.random_.random()` returns a uint64 while
+            // `parent_.dns_jitter_ms_.count()` returns a signed long that gets cast into a uint64.
+            // Thus, the modulo of the two will be a positive as long as
+            // `parent_dns_jitter_ms_.count()` is positive.
+            // It is important that this be positive, otherwise `final_refresh_rate` could be
+            // negative causing Envoy to crash.
+            final_refresh_rate += std::chrono::milliseconds(parent_.random_.random() %
+                                                            parent_.dns_jitter_ms_.count());
+          }
+
+          ENVOY_LOG(debug, "DNS refresh rate reset for {}, refresh rate {} ms", dns_address_,
+                    final_refresh_rate.count());
+        } else {
+          parent_.info_->configUpdateStats().update_failure_.inc();
+
+          final_refresh_rate =
+              std::chrono::milliseconds(parent_.failure_backoff_strategy_->nextBackOffMs());
+          ENVOY_LOG(debug, "DNS refresh rate reset for {}, (failure) refresh rate {} ms",
+                    dns_address_, final_refresh_rate.count());
+        }
+
+        // If there is an initialize callback, fire it now. Note that if the cluster refers to
+        // multiple DNS names, this will return initialized after a single DNS resolution
+        // completes. This is not perfect but is easier to code and unclear if the extra
+        // complexity is needed so will start with this.
+        parent_.onPreInitComplete();
+        resolve_timer_->enableTimer(final_refresh_rate);
+      });
+}
+
+REGISTER_FACTORY(DnsClusterFactory, ClusterFactory);
```

Comment:
I'd suggest to move it closer to the factory definition above and before the DnsClusterImpl